src/main/java/bio/terra/testrunner/runner/TestScriptResult.java

```diff
@@ -38,6 +39,24 @@ private TestScriptResultSummary(String testScriptName, String testScriptDescript
     }
   }
 
+  /** Store snapshots of UserJourneyResults, mainly used for reporting. */
+  @SuppressFBWarnings(
+      value = "URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD",
+      justification = "This POJO class is used for easy serialization to JSON using Jackson.")
+  public static class TestScriptUserJourneySnapshots {
```

Comment:
Ah okay. 2 questions
1. Would it be easier to change the cloud function logic to parse the JSON and insert into BigQuery, rather than modifying the logic here?
2. Can you outline a separate POJO class that has the complete output format you want? (doesn't have to be code, can just be in a comment here). I think that would make it easier to check that we're not duplicating information.

Another related point here. I originally wrote Test Runner library to output a single file. But for runs with lots of metrics readings or hundreds of user journey threads, this single file got unwieldy and started occasionally crashing peoples' text editors. I understand that for automated processing, having things in a single file may make more sense, but I think for developers looking at the files, it would be good to continue writing out separate files. So perhaps we can make this an optional behavior (e.g. continue writing separate files and if a flag is set somewhere, also write the single consolidated file).

Reply:
The cloud function app that we build (in collaboration with App Services) for ingesting JSON files does not require any application-level parsing logic. The JSON is ingested from the Google bucket as raw bytes and `streaming inserts` of the JSON _key-value_ chunks into `BigQuery` will happen at the correct location of the table schema. All we need is a matching table schema with respect to the `TestRunner` output shape.

Here's the table schema of one row of the `TestRunner` summary:

```
[
  {
    "schema": "[\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"id\",\n    \"type\": \"STRING\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"startTime\",\n    \"type\": \"INT64\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"endTime\",\n    \"type\": \"INT64\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"startUserJourneyTime\",\n    \"type\": \"INT64\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"endUserJourneyTime\",\n    \"type\": \"INT64\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"startTimestamp\",\n    \"type\": \"TIMESTAMP\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"endTimestamp\",\n    \"type\": \"TIMESTAMP\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"startUserJourneyTimestamp\",\n    \"type\": \"TIMESTAMP\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"endUserJourneyTimestamp\",\n    \"type\": \"TIMESTAMP\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"testScriptResultSummaries\",\n    \"type\": \"ARRAY<STRUCT<testScriptDescription STRING, testScriptName STRING, elapsedTimeStatistics STRUCT<min FLOAT64, max FLOAT64, mean FLOAT64, standardDeviation FLOAT64, median FLOAT64, percentile95 FLOAT64, percentile99 FLOAT64, sum FLOAT64>, totalRun INT64, numCompleted INT64, numExceptionsThrown INT64, isFailure BOOL>>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"resourceFileName\",\n    \"type\": \"STRING\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"testSuiteName\",\n    \"type\": \"STRING\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"testConfiguration\",\n    \"type\": \"STRUCT<name STRING, description STRING, serverSpecificationFile STRING, billingAccount STRING, isFunctional BOOL, resourceFileName STRING, testUserFiles ARRAY<STRING>>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"server\",\n    \"type\": \"STRUCT<name STRING, description STRING, samUri STRING, datarepoUri STRING, samResourceIdForDatarepo STRING, bufferUri STRING, bufferClientServiceAccountFile STRING, bufferClientServiceAccount STRUCT<name STRING, jsonKeyDirectoryPath STRING, jsonKeyFilename STRING>>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"kubernetes\",\n    \"type\": \"STRUCT<numberOfInitialPods INT64>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"application\",\n    \"type\": \"STRUCT<maxStairwayThreads INT64, maxBulkFileLoad INT64, maxBulkFileLoadArray INT64, loadConcurrentFiles INT64, loadConcurrentIngests INT64, loadDriverWaitSeconds INT64, loadHistoryCopyChunkSize INT64, loadHistoryWaitSeconds INT64>\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"testScripts\",\n    \"type\": \"ARRAY<STRUCT<name STRING, numberOfUserJourneyThreadsToRun INT64, userJourneyThreadPoolSize INT64, expectedTimeForEach INT64, expectedTimeForEachUnit STRING, parameters ARRAY<STRING>, expectedTimeForEachUnitObj STRING, description STRING>>\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"testUsers\",\n    \"type\": \"ARRAY<STRUCT<name STRING, userEmail STRING, delegatorServiceAccountFile STRING, delegatorServiceAccount STRUCT<name STRING, jsonKeyDirectoryPath STRING, jsonKeyFilename STRING>>>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"disruptiveScript\",\n    \"type\": \"STRUCT<name STRING, parameters ARRAY<STRING>>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"testConfig\",\n    \"type\": \"STRUCT<name STRING, description STRING, serverSpecificationFile STRING, billingAccount STRING, isFunctional BOOL, resourceFileName STRING, testUserFiles ARRAY<STRING>, server STRUCT<name STRING, description STRING, samUri STRING, datarepoUri STRING, samResourceIdForDatarepo STRING, bufferUri STRING, bufferClientServiceAccountFile STRING, bufferClientServiceAccount STRUCT<name STRING, jsonKeyDirectoryPath STRING, jsonKeyFilename STRING>, workspaceManagerUri STRING, policyManagerUri STRING, cluster STRUCT<clusterName STRING, clusterShortName STRING, region STRING, zone STRING, project STRING, namespace STRING, containerName STRING, componentLabel STRING, apiComponentLabel STRING>, testRunnerServiceAccountFile STRING, testRunnerServiceAccount STRUCT<name STRING, jsonKeyDirectoryPath STRING, jsonKeyFilename STRING>, testRunnerK8SServiceAccountFile STRING, testRunnerK8SServiceAccount STRUCT<name STRING, clientKeyDirectoryPath STRING, clientKeyFilename STRING, tokenFilename STRING>, deploymentScript STRUCT<name STRING, parameters ARRAY<STRING>, scriptClass STRING>, skipKubernetes BOOL, skipDeployment BOOL>, kubernetes STRUCT<numberOfInitialPods INT64>, application STRUCT<maxStairwayThreads INT64, maxBulkFileLoad INT64, maxBulkFileLoadArray INT64, loadConcurrentFiles INT64, loadConcurrentIngests INT64, loadDriverWaitSeconds INT64, loadHistoryCopyChunkSize INT64, loadHistoryWaitSeconds INT64>, testScripts ARRAY<STRUCT<name STRING, numberOfUserJourneyThreadsToRun INT64, userJourneyThreadPoolSize INT64, expectedTimeForEach INT64, expectedTimeForEachUnit STRING, parameters ARRAY<STRING>, expectedTimeForEachUnitObj STRING, description STRING>>, testUsers ARRAY<STRUCT<name STRING, userEmail STRING, delegatorServiceAccountFile STRING, delegatorServiceAccount STRUCT<name STRING, jsonKeyDirectoryPath STRING, jsonKeyFilename STRING>>>, disruptiveScript STRUCT<name STRING, parameters ARRAY<STRING>>>\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"userJourneySnapshotsCollection\",\n    \"type\": \"ARRAY<STRUCT<testScriptName STRING, testScriptDescription STRING, userJourneyResults ARRAY<STRUCT<userJourneyDescription STRING, threadName STRING, completed BOOL, elapsedTimeNS INT64, exceptionWasThrown BOOL, exceptionStackTrace STRING, exceptionMessage STRING>>>>\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"terraVersions\",\n    \"type\": \"ARRAY<STRUCT<appName STRING, appVersion STRING, chartVersion STRING>>\"\n  }\n]"
  }
]
```

Reply:
You are right on point that this is meant for automated processing, not something I would expect to be opened in a text editor for inspection. The `BigQuery streaming insert` provides a generous rate limit (up to 10MB per row) and 1GB per second without using `insertId` which is more than sufficient for `TestRunner` across all `MC Terra`. The consolidated summary is <= 10k bytes for `WSM` for each test suite in nightly run.

Reply:
The cloud function only process `SUMMARY_testRun.json`. Other files like `RAWDATA_userJourneyResults.json` and `RENDERED_testConfiguration.json` still exist in the uploaded `tgz` and can be downloaded. 

Instead of having a switch to get the original shorter `SUMMARY_testRun.json`. How about add a `SUMMARY_testRun_short.json` to the `tgz` as a additional file?

Reply:
I like that idea. Slight modification, can you leave the existing filenames in case someone depends on them, and write a new one e.g. `SUMMARY_testRun_concatenated.json`?

Reply:
As part of this work, can you pull out the `TestRunSummary` and your new output file JSON format into separate POJO classes? I think it's easier to read if we have all the file formats in their own classes. e.g. for your new format:

```
public static class TestRunSummaryConcatenated {
  public TestRunSummary summary;
  public TestConfiguration config;
  public List<TestScriptResult> testScriptResults;
}
```

Reply:
> I like that idea. Slight modification, can you leave the existing filenames in case someone depends on them, and write a new one e.g. `SUMMARY_testRun_concatenated.json`?

Yea that's what I intended to comment on but ended up not correcting it when finalizing my comments last Friday. Yes, what I should have said is to keep the original `SUMMARY_testRun.json` and use a new name for the longer version like what you've suggested `SUMMARY_testRun_concatenated.json`.

Reply:
> As part of this work, can you pull out the `TestRunSummary` and your new output file JSON format into separate POJO classes? I think it's easier to read if we have all the file formats in their own classes. e.g. for your new format:
> 
> ```
> public static class TestRunSummaryConcatenated {
>   public TestRunSummary summary;
>   public TestConfiguration config;
>   public List<TestScriptResult> testScriptResults;
> }
> ```

I did it slightly differently. Instead I created a subclass of `TestRunSummary` 

```
public static class TestRunSummaryConcatenated extends TestRunSummary
```

and use the following views  (@JsonView(SummaryViews.Summary.class) and @JsonView(SummaryViews.ConcatenatedSummary.class)) to generate `SUMMARY_testRun.json` and `SUMMARY_concat_testRun.json` (prefer a shorter name compared to `SUMMARY_testRun_concatenated.json`).

```
public class SummaryViews {
  /** JSON View for TestRunSummary. */
  public static class Summary {}

  /** JSON view for TestRunSummaryConcatenated. */
  public static class ConcatenatedSummary extends Summary {}
}
```